### PR TITLE
Remove `on_device_shape` and `logical_on_device_shape`

### PR DESCRIPTION
**Pull request by will-cromar**
These were removed from the PJRT C API client recently in this commit: https://github.com/tensorflow/tensorflow/commit/840633cb23ae111075999514484fd5416f0cef83

Instead, we can rebuild the shapes from the decomposed components. In the future, we should see if we can remove `Shape` from `ComputationClient::Data` entirely in favor of `dimensions`, `layout`, etc.

### CODE REVIEW

**Review by alanwaketan — APPROVED** [review]
LGTM.

**Review by yeounoh — APPROVED** [review]
LGTM

**Comment by will-cromar** [comment]
Once the openXLA change to add `is_dynamic_dimension` is submitted, I'll add the commit hash to the patch in this PR and mark it ready to submit.

**Review by JackCaoG — APPROVED** [review]

**Review by vanbasten23 — APPROVED** [review]

**Comment by will-cromar** [comment]
The upstream commit merged: https://github.com/openxla/xla/commit/6308dba2903e78961ac4122f361bc91b09f36891

Added the hash to the patch in this commit. Let's wait until after we're done moving the OpenXLA pin to merge this (see #5592)

**Comment by ManfeiBai** [comment]
Do we want this PR in 2.1 branch?

**Comment by will-cromar** [comment]
No. This is a risky change, so we should not backport it.

**Pull request merged**
